crates/sound/src/stream.rs

```diff
@@ -0,0 +1,305 @@
+///
+///   Possible valid transitions: `START`, `RELEASE`.
+///
+/// - `RELEASE`
+///
+///   The device releases the stream (frees resources, etc).
+///
+///   Possible valid transitions: `SET PARAMETERS`, `PREPARE`.
+///
+/// ```text
+/// +---------------+ +---------+ +---------+ +-------+ +-------+
+/// | SetParameters | | Prepare | | Release | | Start | | Stop  |
+/// +---------------+ +---------+ +---------+ +-------+ +-------+
+///         |              |           |          |         |
+///         |-             |           |          |         |
+///         ||             |           |          |         |
+///         |<             |           |          |         |
+///         |              |           |          |         |
+///         |------------->|           |          |         |
+///         |              |           |          |         |
+///         |<-------------|           |          |         |
+///         |              |           |          |         |
+///         |              |-          |          |         |
+///         |              ||          |          |         |
+///         |              |<          |          |         |
+///         |              |           |          |         |
+///         |              |--------------------->|         |
+///         |              |           |          |         |
+///         |              |---------->|          |         |
+///         |              |           |          |         |
+///         |              |           |          |-------->|
+///         |              |           |          |         |
+///         |              |           |          |<--------|
+///         |              |           |          |         |
+///         |              |           |<-------------------|
+///         |              |           |          |         |
+///         |<-------------------------|          |         |
+///         |              |           |          |         |
+///         |              |<----------|          |         |
+/// ```
+#[derive(Debug, Default, Copy, Clone)]
+pub enum PCMState {
+    #[default]
+    #[doc(alias = "VIRTIO_SND_R_PCM_SET_PARAMS")]
+    SetParameters,
+    #[doc(alias = "VIRTIO_SND_R_PCM_PREPARE")]
+    Prepare,
+    #[doc(alias = "VIRTIO_SND_R_PCM_RELEASE")]
+    Release,
+    #[doc(alias = "VIRTIO_SND_R_PCM_START")]
+    Start,
+    #[doc(alias = "VIRTIO_SND_R_PCM_STOP")]
+    Stop,
+}
+
+macro_rules! set_new_state {
+    ($new_state_fn:ident, $new_state:expr, $($valid_source_states:tt)*) => {
+        pub fn $new_state_fn(&mut self) -> Result<()> {
+            if !matches!(self, $($valid_source_states)*) {
+                return Err(Error::InvalidStateTransition(*self, $new_state));
+            }
+            *self = $new_state;
+            Ok(())
+        }
+    };
+}
+
+impl PCMState {
+    pub fn new() -> Self {
+        Self::SetParameters
+    }
```

Comment:
Sounds like a Default impl ;)

Reply:
forgot to comment, fixed in latest force push *